paper/automated.tex

```diff
@@ -1,20 +1,108 @@
+\label{sec:proof-reconstruction}
+
+While ATPs are very useful to solve questions for this project, they generally don't integrate well with Lean (except for Lean tactics like \texttt{duper}, \texttt{aesop} and \texttt{egg}).
+A bug in an ATP could lead to an unsound proof, or worse, an incorrect result.
+To avoid this, and having to trust large codebases from ATPs, we take results found by the different provers and (re-)construct proofs in Lean.
+
+Integration implies two steps: invoking the decision procedures/ATPs (translating the problem from Lean into the languages and logics they use), and conversely, (re-)constructing the results from the decision procedure as a (persistent) Lean proof.
+These two aspect present different challenges, and require different strategies, depending mostly on the kind of proof strategy the decision procedure uses.
+
+For saturation proofs $\ldots$ (TODO: add explanations from \url{https://leanprover.zulipchat.com/#narrow/channel/458659-Equational/topic/Vampire.20-.3E.20Lean/near/478147138})
+
+For equational proofs from external provers (e.g. MagmaEgg), we used also a simple version of reconstruction by (re-)constructing the proofs of equality from an explanation, using congruence lemmas from Lean.
+In the case of equational proofs from the \texttt{egg} tactic, these could be converted into a series of \texttt{calc} steps, documenting the explicit calculation, using the \texttt{calcify} tactic\footnote{\url{https://github.com/nomeata/lean-calcify/}}.
+
+In general, we've observed that there are multiple ways of integrating decisions procedures within Lean, with different levels of integration.
+\begin{enumerate}
+    \item Using a Lean tactic, which calls a decision procedure written in Lean (like \texttt{aesop} or \texttt{duper}).
+    \item Using a Lean tactic, which calls an existing (external) ATP and reconstructs a proof term from the ATP's result (like \texttt{bv\_decide} or \texttt{egg}).
+    \item\label{external} Using an external script which calls an existing ATP and generates a source file \texttt{.lean} which captures the result explicitly.
+\end{enumerate}
+
+
+This project primarily used the least integrated approach, (Option \ref{external}), as it was fastest and required no dependencies on the other contributors.
+This also has drawbacks: primarily, while the upfront effort is lower, the effort to use is higher than with a tactic, once the tactic is developed.
+It also makes integration with larger code bases more difficult: in this project the (mathematical) dependencies were by design very minimal, Magmas are simple, and we built our own definitions for them.
+For example, integration with the typeclass system become much more difficult when working with more complex mathematical objects that build upon many layers of structure.
+
+\paragraph{Semi-Automated Counterexample Guidance}
+
+Another use of ATPs has been in a semi-automatic fashion, to find counter-examples.
```

Comment:
I would maybe get rid of "in a semi-automatic fashion," here. There are examples of fully-automated and quite bespoke counterexample construction. It may be interesting to mention that, or leave a TODO for the future.

Reply:
I don't understand the issue with that, but I have also been less involved in this effort. Semi-automated subsumes fully automated, and not all have been fully automated, right? More importantly though, the entire section has talked about fully automated approaches, this is just saying it's not *only* been fully automated, there have been semi-automated approaches. Maybe this should be expanded in a different PR after?